### PR TITLE
refactor: make statistics a parameter of DatafusionOptimizer

### PR DESCRIPTION
**Pull request by Gun9niR**
Add base table statistics as a parameter in `DatafusionOptimizer`.

### CODE REVIEW

File: datafusion-optd-cli/src/main.rs [hunk 1]
```diff
@@ -203,15 +204,15 @@ pub async fn main() -> Result<()> {
...
-                state.catalog_list(),
-            )))
-        };
+        let runtime_statistics = RuntimeAdaptionStorage::default();
+        let cost_model =
+            AdaptiveCostModel::new(DEFAULT_DECAY, runtime_statistics.clone(), Stats::default());
+        let optimizer = DatafusionOptimizer::new_physical(
+            Arc::new(DatafusionCatalog::new(state.catalog_list())),
+            Box::new(cost_model),
+            runtime_statistics,
```
Comment (patrickhyw) [inline datafusion-optd-cli/src/main.rs]: If the stats are already stored in the cost model and the cost model is sent to the optimizer, why do we send another copy of the stats to the optimizer? Can't the optimizer just use the stats stored in the cost model?

File: optd-adaptive-demo/src/bin/optd-adaptive-three-join.rs [hunk 2]
```diff
@@ -27,8 +28,14 @@ async fn main() -> Result<()> {
         let runtime_env = RuntimeEnv::new(rn_config.clone())?;
         let mut state =
             SessionState::new_with_config_rt(session_config.clone(), Arc::new(runtime_env));
-        let mut optimizer: DatafusionOptimizer = DatafusionOptimizer::new_physical_adaptive(
+        let runtime_statistics = RuntimeAdaptionStorage::default();
+        let cost_model =
+            AdaptiveCostModel::new(DEFAULT_DECAY, runtime_statistics.clone(), Stats::default());
```
Comment (wangpatrick57) [inline optd-adaptive-demo/src/bin/optd-adaptive-three-join.rs:33]: Same comment about stats.clone() as above

File: optd-adaptive-demo/src/bin/optd-adaptive-tpch-q8.rs [hunk 3]
```diff
@@ -28,9 +29,15 @@ async fn main() -> Result<()> {
     let mut ctx = {
         let mut state =
             SessionState::new_with_config_rt(session_config.clone(), Arc::new(runtime_env));
-        let optimizer = DatafusionOptimizer::new_physical_adaptive(Arc::new(
-            DatafusionCatalog::new(state.catalog_list()),
-        ));
+        let runtime_statistics = RuntimeAdaptionStorage::default();
+        let cost_model =
+            AdaptiveCostModel::new(DEFAULT_DECAY, runtime_statistics.clone(), Stats::default());
+        let optimizer: DatafusionOptimizer = DatafusionOptimizer::new_physical(
```
Comment (wangpatrick57) [inline optd-adaptive-demo/src/bin/optd-adaptive-tpch-q8.rs:35]: Same comment

File: optd-sqlplannertest/src/lib.rs [hunk 4]
```diff
@@ -61,9 +62,15 @@ impl DatafusionDb {
             } else {
                 SessionState::new_with_config_rt(session_config.clone(), Arc::new(runtime_env))
             };
-            let optimizer = DatafusionOptimizer::new_physical(Arc::new(DatafusionCatalog::new(
-                state.catalog_list(),
-            )));
+            let runtime_statistics = RuntimeAdaptionStorage::default();
+            let cost_model =
+                AdaptiveCostModel::new(DEFAULT_DECAY, runtime_statistics.clone(), Stats::default());
```
Comment (wangpatrick57) [inline optd-sqlplannertest/src/lib.rs:67]: Same comment

File: optd-perftest/src/datafusion_db_cardtest.rs [hunk 5]
```diff
@@ -82,9 +85,15 @@ impl DatafusionDb {
         let ctx = {
             let mut state =
                 SessionState::new_with_config_rt(session_config.clone(), Arc::new(runtime_env));
-            let optimizer = DatafusionOptimizer::new_physical(Arc::new(DatafusionCatalog::new(
-                state.catalog_list(),
-            )));
+            let runtime_statistics = RuntimeAdaptionStorage::default();
+            let cost_model =
+                AdaptiveCostModel::new(DEFAULT_DECAY, runtime_statistics.clone(), Stats::default());
```
Comment (wangpatrick57) [inline optd-perftest/src/datafusion_db_cardtest.rs:90]: Same comment

**Review by patrickhyw — COMMENTED** [review]
LGTM other than my comment

**Pull request merged**